### PR TITLE
Memory observers

### PR DESCRIPTION
**Pull request by dedecun**
### Why was it necessary?

In the mobile environment the codelets should run only when the memory input changes, so an observer pattern was implemented for this control.

### CODE REVIEW

File: src/main/java/br/unicamp/cst/core/entities/Memory.java [hunk 1]
```diff
@@ -79,5 +79,11 @@
 	 */
 	public Long getTimestamp();
 	
+	/**
+	 * Add a memory observer to its list
+	 * @param memoryObserver
+	 */
+	public void addMemoryObservers(MemoryObserver memoryObserver);
```
Comment (wandemberg-eldorado) [inline src/main/java/br/unicamp/cst/core/entities/Memory.java:86]: please remove the 's' as this method only adds a single memoryObserver

```suggestion
	public void addMemoryObserver(MemoryObserver memoryObserver);
```

File: src/test/java/br/unicamp/cst/core/entities/MemoryTest.java [hunk 2]
```diff
@@ -53,6 +53,12 @@ public void setEvaluation(Double eval) {
         public Long getTimestamp() {
             return this.timestamp;
         }
+
+		@Override
+		public void addMemoryObservers(MemoryObserver memoryObserver) {
```
Comment (wandemberg-eldorado) [inline src/test/java/br/unicamp/cst/core/entities/MemoryTest.java:58]:
```suggestion
		public void addMemoryObserver(MemoryObserver memoryObserver) {
```

File: src/test/java/br/unicamp/cst/util/TestMemoryObserver.java [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+	public void proc() {
+		counter++;
+	}
+
+}
+
+public class TestMemoryObserver {
+
+	@Test
+	public void test1() throws InterruptedException {
```
Comment (wandemberg-eldorado) [inline src/test/java/br/unicamp/cst/util/TestMemoryObserver.java:47]: please avoid generic names on tests as we would avoid it in other methods

```suggestion
	public void noMemoryChangeTest() throws InterruptedException {
```

File: src/test/java/br/unicamp/cst/util/TestMemoryObserver.java [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+		m.start();
+		Thread.sleep(2000);
+		m.shutDown();
+
+		assertEquals(1, c.getCounter());
+		assertEquals(1, c2.getCounter());
+	}
+
+	@Test
+	public void test2() throws InterruptedException {
```
Comment (wandemberg-eldorado) [inline src/test/java/br/unicamp/cst/util/TestMemoryObserver.java:97]:
```suggestion
	public void usualRunTest() throws InterruptedException {
```

File: src/test/java/br/unicamp/cst/util/TestMemoryObserver.java [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+		m8.setI(100, 0);
+		Thread.sleep(2000);
+		m.shutDown();
+
+		assertEquals(2, c.getCounter());
+		assertEquals(3, c2.getCounter());
+	}
+	
+	@Test
+	public void test3() throws InterruptedException {
```
Comment (wandemberg-eldorado) [inline src/test/java/br/unicamp/cst/util/TestMemoryObserver.java:153]:
```suggestion
	public void multipleMemoryChangesTest() throws InterruptedException {
```

File: src/test/java/br/unicamp/cst/util/TestMemoryObserver.java [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+		m4.setI(2);
+		Thread.sleep(2000);
+		m.shutDown();
+
+		assertEquals(5, c.getCounter());
+		assertEquals(2, c2.getCounter());
+	}
+	
+	@Test
+	public void test4() throws InterruptedException {
```
Comment (wandemberg-eldorado) [inline src/test/java/br/unicamp/cst/util/TestMemoryObserver.java:207]:
```suggestion
	public void addSameMemoryTest() throws InterruptedException {
```

**Review by wandemberg-eldorado — COMMENTED** [review]
minor name correction.
Test name suggestions

**Review by wandemberg-eldorado — APPROVED** [review]

**Review by andre-paraense — APPROVED** [review]

**Pull request merged**